pkg/csi_driver/multishare_ops_manager.go

```diff
@@ -280,327 +225,300 @@ func (m *MultishareOpsManager) startShareWorkflow(ctx context.Context, instanceS
-			}
+	ops, err := m.listMultishareResourceRunningOps(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
 
-			if shareName == targetShareName {
-				return share, nil
-			}
-		}
+	// If we find a running delete share op, poll for that to complete.
+	deleteshareop, err := containsOpWithShareTarget(share, util.ShareDelete, ops)
+	if err != nil {
+		return nil, err
+	}
+	if deleteshareop != nil {
+		return &Workflow{share: share, opName: deleteshareop.Id, opType: deleteshareop.Type}, nil
 	}
 
-	return nil, nil
+	return m.startShareWorkflow(ctx, &Workflow{share: share, opType: util.ShareDelete}, ops)
 }
 
-// runEligibleInstanceCheck returns a list of ready and non-ready instances.
-func (m *MultishareOpsManager) runEligibleInstanceCheck(ctx context.Context, instanceScPrefix string) ([]*file.MultishareInstance, int, error) {
-	var readyInstanceKeys []util.InstanceKey
-	nonReadyInstanceCount := 0
-	items := m.cache.GetInstanceMap(instanceScPrefix).Items()
-	for _, item := range items {
-		instanceKey := item.Key
-		opInfo := item.OpInfo
-		project, location, instanceName, err := util.ParseInstanceHandle(string(instanceKey))
-		if err != nil {
-			klog.Warningf("Failed to parse instance key %v", instanceKey)
-			continue
-		}
-		if opInfo.Name == "" {
-			readyInstanceKeys = append(readyInstanceKeys, instanceKey)
-			continue
-		}
+func (m *MultishareOpsManager) checkAndStartInstanceDeleteOrShrinkWorkflow(ctx context.Context, instance *file.MultishareInstance) (*Workflow, error) {
+	m.Lock()
+	defer m.Unlock()
 
-		// check and clear completed op
-		ready, err := m.verifyNoRunningInstanceOps(ctx, instanceScPrefix, &file.MultishareInstance{
-			Project:  project,
-			Location: location,
-			Name:     instanceName,
-		})
-		if err != nil {
-			klog.Warningf("Failed to check instance ready for instance key %v, err:%v", instanceKey, err)
-		}
+	ops, err := m.listMultishareResourceRunningOps(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	err = m.verifyNoRunningInstanceOrShareOpsForInstance(instance, ops)
+	if err != nil {
+		return nil, err
+	}
 
-		if !ready && opInfo.Type != util.InstanceDelete {
-			nonReadyInstanceCount = nonReadyInstanceCount + 1
+	// At this point no new share create or delete would be attempted since the driver has the lock.
+	// 1. GET instance . if not found its a no-op return success.
+	// 2. evaluate 0 shares.
+	// 3. else evaluate instance size with share size sum.
+	instance, err = m.cloud.File.GetMultishareInstance(ctx, instance)
+	if err != nil {
+		if file.IsNotFoundErr(err) {
+			return nil, nil
 		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
 
-		if ready {
-			readyInstanceKeys = append(readyInstanceKeys, instanceKey)
+	shares, err := m.cloud.File.ListShares(ctx, &file.ListFilter{Project: instance.Project, Location: instance.Location, InstanceName: instance.Name})
+	if err != nil {
+		if file.IsNotFoundErr(err) {
```

Comment:
if there's not found error, it should mean there's no share on this instance IIUC? so we should delete the instance instead of returning right?

Reply:
As discussed, not found would be returned if the instance is missing. empty share list returns [], and nil error.